docs/content/oauth-oidc.md

```diff
@@ -0,0 +1,131 @@
+
+## Token issuers, discovery and metadata
+
+Any JWT issued by an OAuth2 or OpenID Connect authorization server will carry the URL of the issuer as part of the `iss` claim in the token payload. This value _must_ be matched against the expected issuer(s) when verifying the token. OpenID Connect standardized on requiring the issuer to be a valid HTTPS URL pointing to the address of the issuer. This URL can be used for discovery purposes, obtaining information from the issuer, like available endpoints, that would otherwise need to be kept in applications and policies. The [metadata endpoint](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata) of any compliant OpenID Connect server is found by appending the `/.well-known/openid-configuration` endpoint to the issuer URL. In order to reduce the number of configuration options coded into our policies, we may query this URL and use the endpoints provided in the response for further interactions with the authorization server.
+
+```live:oidc:module
+package oidc
+
+# Error handling and JWT verification omitted for brevity
+
+issuers = {"https://issuer1.example.com", "https://issuer2.example.com"}
+
+metadata(issuer) = http.send({
+    "url": concat("", issuers[issuer], "/.well-known/openid-configuration",
+    "method": "GET",
+    "force_cache": true,
+    "force_cache_duration_seconds": 86400 # Cache response for 24 hours
+}).body
+
+claims := jwt.decode(input.token)[1]
+
+jwks_endpoint := metadata(claims.iss).jwks_uri
+token_endpoint := metadata(claims.iss).token_endpoint
+```
+While not originally part of the OAuth2 specification, the metadata specification from OpenID Connect has later been ["backported" to OAuth2](https://tools.ietf.org/html/rfc8414). Since all endpoints are practically the same, it is however common to only use the OpenID Connect metadata endpoint where available.
+
+## JWKS endpoint and token verification
+
+A [JSON Web Key Set](https://tools.ietf.org/html/rfc7517) is a collection of cryptographic keys in JSON format. The public key components of these are used to verify the authenticity of JWTs and can be provided as is to the JWT verification functions of OPA. While these could be pulled into OPA as part of data bundles or by other means provided to a policy, the preferred way is to query the authorization server directly. Standards compliant token issuers will publish the public key component of any signature keys it uses at it's JWKS endpoint (`jwks_uri` from the metadata endpoint). Any key published there will have a key ID (`kid`), which will also be present in the header of any JWT signed by the issuer. Using OPA's `http.send` function we can turn directly to the JWKS endpoint of the issuer to retrieve the public keys needed for token verification.
+
+```live:oidc2:module
+package oidc
+
+# Error handling omitted for brevity
+
```

Comment:
Great content, we use something similar. But our jwks_uri move around so for us to be safe we actually end up with two requests. Perhaps the 2nd example should reflect this?
`
json_jwks = r {
	not jwt_token_kid_exist_in_local_keyset
	openid_request_body = {
		"method": "GET",
		"url": sprintf("%v/.well-known/openid-configuration", [config.issuer]),
		"force_cache": true,
		"force_cache_duration_seconds": 3600,
	}

	openid_response = http.send(openid_request_body)
	jwks_request_body = {
		"method": "GET",
		"url": openid_response.body.jwks_uri,
		"force_cache": true,
		"force_cache_duration_seconds": 3600,
	}

	jwks_response = http.send(jwks_request_body)
	r = jwks_response.raw_body
}
`

Reply:
You mean I should practice what I preach? Get out of here @johanneslarsson ! 😄 

I wanted the examples to be "standalone" so to speak, but I guess you are right - since discovery is explained early on we might as well use it in subsequent examples too. I'll give it some thought.

The `jwks_uri` moving around has me intrigued. Never encountered that before - but using discovery certainly makes it possible. What vendor/product does that?

Reply:
Pretty much, yes 😆 . It's not that frequent. But after spending hours on correcting the hardcoded 'jwks_uri' you realise it's better to follow the standard.